activesupport/lib/active_support/lazy_load_hooks.rb

```diff
@@ -27,33 +27,60 @@ def self.extended(base) # :nodoc:
       base.class_eval do
         @load_hooks = Hash.new { |h, k| h[k] = [] }
         @loaded     = Hash.new { |h, k| h[k] = [] }
+        @only_once  = Hash.new { |h, k| h[k] = [] }
       end
     end
 
     # Declares a block that will be executed when a Rails component is fully
     # loaded.
+    #
+    # Options:
+    #
+    # * <tt>:yield</tt> - Yields the object that run_load_hooks to +block+.
+    # * <tt>:only_once</tt> - Given +block+ will run only once.
     def on_load(name, options = {}, &block)
       @loaded[name].each do |base|
-        execute_hook(base, options, block)
+        process_hook(name, base, options, block)
       end
 
       @load_hooks[name] << [block, options]
     end
 
-    def execute_hook(base, options, block)
-      if options[:yield]
-        block.call(base)
-      else
-        base.instance_eval(&block)
-      end
-    end
-
     def run_load_hooks(name, base = Object)
       @loaded[name] << base
       @load_hooks[name].each do |hook, options|
-        execute_hook(base, options, hook)
+        process_hook(name, base, options, hook)
       end
     end
+
+    private
+
+      # Executes a hook. If the hook was loaded providing :only_once option,
+      # it is flagged after the first time it is executed and then skipped.
+      def process_hook(name, base, options, block)
+        return if already_run?(name, block)
+
+        execute_hook(base, options, block)
+        flag_run(name, options, block)
+      end
+
+      def execute_hook(base, options, block)
+        if options[:yield]
+          block.call(base)
+        else
+          base.instance_eval(&block)
+        end
+      end
+
+      def already_run?(name, hook)
+        @only_once[name].include?(hook)
+      end
+
+      def flag_run(name, options, hook)
+        if options[:only_once]
+          @only_once[name] << hook
+        end
+      end
```

Comment:
This seems a little too abstracted for my taste. I think it would be clearer if we kept the `once` checking responsibility within the same method:

```ruby
def execute_hook(name, base, options, block)
  with_once_execution(name, block, options[:only_once]) do
    if options[:yield]
      block.call(base)
    else
      base.instance_eval(&block)
    end
  end
end

def with_once_execution(name, block, once)
  unless @only_once[name].include?(block)
    @only_once[name] << block if once

    yield
  end
end
```

(I'd probably just call the option `once: true` since the `only_` is already implied.)